homeassistant/components/alarmdecoder/config_flow.py

```diff
@@ -0,0 +1,322 @@
+from homeassistant.core import callback
+
+from .const import (  # pylint: disable=unused-import
+    CONF_ALT_NIGHT_MODE,
+    CONF_AUTO_BYPASS,
+    CONF_CODE_ARM_REQUIRED,
+    CONF_DEVICE_BAUD,
+    CONF_DEVICE_PATH,
+    CONF_RELAY_ADDR,
+    CONF_RELAY_CHAN,
+    CONF_ZONE_LOOP,
+    CONF_ZONE_NAME,
+    CONF_ZONE_NUMBER,
+    CONF_ZONE_RFID,
+    CONF_ZONE_TYPE,
+    DEFAULT_ARM_OPTIONS,
+    DEFAULT_DEVICE_BAUD,
+    DEFAULT_DEVICE_HOST,
+    DEFAULT_DEVICE_PATH,
+    DEFAULT_DEVICE_PORT,
+    DEFAULT_ZONE_OPTIONS,
+    DEFAULT_ZONE_TYPE,
+    DOMAIN,
+    OPTIONS_ARM,
+    OPTIONS_ZONES,
+    PROTOCOL_SERIAL,
+    PROTOCOL_SOCKET,
+)
+
+EDIT_KEY = "edit_selection"
+EDIT_ZONES = "Zones"
+EDIT_SETTINGS = "Arming Settings"
+
+
+class AlarmDecoderFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a AlarmDecoder config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH
+
+    def __init__(self):
+        """Initialize AlarmDecoder ConfigFlow."""
+        self.protocol = None
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(config_entry):
+        """Get the options flow for AlarmDecoder."""
+        return AlarmDecoderOptionsFlowHandler(config_entry)
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initialized by the user."""
+        if self._async_current_entries():
+            return self.async_abort(reason="single_instance_allowed")
+
+        if user_input is not None:
+            self.protocol = user_input[CONF_PROTOCOL]
+            return await self.async_step_protocol()
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_PROTOCOL): vol.In(
+                        [PROTOCOL_SOCKET, PROTOCOL_SERIAL]
+                    ),
+                }
+            ),
+        )
+
+    async def async_step_protocol(self, user_input=None):
+        """Handle AlarmDecoder protocol setup."""
+        errors = {}
+        if user_input is not None:
+            connection = {}
+            if self.protocol == PROTOCOL_SOCKET:
+                baud = connection[CONF_DEVICE_BAUD] = DEFAULT_DEVICE_BAUD
+                host = connection[CONF_HOST] = user_input[CONF_HOST]
+                port = connection[CONF_PORT] = user_input[CONF_PORT]
+                title = f"{host}:{port}"
+                device = SocketDevice(interface=(host, port))
+            if self.protocol == PROTOCOL_SERIAL:
+                path = connection[CONF_DEVICE_PATH] = user_input[CONF_DEVICE_PATH]
+                baud = connection[CONF_DEVICE_BAUD] = user_input[CONF_DEVICE_BAUD]
+                title = path
+                device = SerialDevice(interface=path)
+
+            controller = AdExt(device)
+            try:
+                with controller:
+                    controller.open(baudrate=baud)
```

Comment:
Does the connection need to be closed here ?

Reply:
the connection should close after exiting the context manager ([src](https://github.com/nutechsoftware/alarmdecoder/blob/master/alarmdecoder/devices/base_device.py#L60-L66))

Reply:
Is this async safe?